File: project/.sdk/tm/py/core/context.py
# ProjectName SDK context

from __future__ import annotations
import random

from utility.voxgig_struct import voxgig_struct as vs
from core.control import ProjectNameControl
from core.operation import ProjectNameOperation
from core.spec import ProjectNameSpec
from core.result import ProjectNameResult
from core.response import ProjectNameResponse
from core.error import ProjectNameError
from core.helpers import get_ctx_prop, to_map


class ProjectNameContext:
    def __init__(self, ctxmap=None, basectx=None):
        self.id = "C" + str(random.randint(10000000, 99999999))
        self.out = {}

        if ctxmap is None:
            ctxmap = {}

        # Client
        c = get_ctx_prop(ctxmap, "client")
        if c is not None:
            self.client = c
        elif basectx is not None:
            self.client = basectx.client
        else:
            self.client = None

        # Utility
        u = get_ctx_prop(ctxmap, "utility")
        if u is not None:
            self.utility = u
        elif basectx is not None:
            self.utility = basectx.utility
        else:
            self.utility = None

        # Ctrl
        self.ctrl = ProjectNameControl()
        ctrl_raw = get_ctx_prop(ctxmap, "ctrl")
        if isinstance(ctrl_raw, dict):
            if ctrl_raw.get("throw_err") is not None:
                self.ctrl.throw_err = ctrl_raw["throw_err"]
            elif isinstance(ctrl_raw.get("throw"), bool):
                self.ctrl.throw_err = ctrl_raw["throw"]
            if isinstance(ctrl_raw.get("explain"), dict):
                self.ctrl.explain = ctrl_raw["explain"]
        elif basectx is not None and basectx.ctrl is not None:
            self.ctrl = basectx.ctrl

        # Meta
        self.meta = {}
        m = get_ctx_prop(ctxmap, "meta")
        if isinstance(m, dict):
            self.meta = m
        elif basectx is not None and basectx.meta is not None:
            self.meta = basectx.meta

        # Config
        cfg = get_ctx_prop(ctxmap, "config")
        if isinstance(cfg, dict):
            self.config = cfg
        elif basectx is not None:
            self.config = basectx.config
        else:
            self.config = None

        # Entopts
        eo = get_ctx_prop(ctxmap, "entopts")
        if isinstance(eo, dict):
            self.entopts = eo
        elif basectx is not None:
            self.entopts = basectx.entopts
        else:
            self.entopts = None

        # Options
        o = get_ctx_prop(ctxmap, "options")
        if isinstance(o, dict):
            self.options = o
        elif basectx is not None:
            self.options = basectx.options
        else:
            self.options = None

        # Entity
        e = get_ctx_prop(ctxmap, "entity")
        if e is not None:
            self.entity = e
        elif basectx is not None:
            self.entity = basectx.entity
        else:
            self.entity = None

        # Shared
        s = get_ctx_prop(ctxmap, "shared")
        if isinstance(s, dict):
            self.shared = s
        elif basectx is not None:
            self.shared = basectx.shared
        else:
            self.shared = None

        # Opmap
        om = get_ctx_prop(ctxmap, "opmap")
        if isinstance(om, dict):
            self.opmap = om
        elif basectx is not None:
            self.opmap = basectx.opmap
        else:
            self.opmap = None
        if self.opmap is None:
            self.opmap = {}

        # Data
        self.data = to_map(get_ctx_prop(ctxmap, "data")) or {}
        self.reqdata = to_map(get_ctx_prop(ctxmap, "reqdata")) or {}
        self.match = to_map(get_ctx_prop(ctxmap, "match")) or {}
        self.reqmatch = to_map(get_ctx_prop(ctxmap, "reqmatch")) or {}

        # Point
        pt = get_ctx_prop(ctxmap, "point")
        if isinstance(pt, dict):
            self.point = pt
        elif basectx is not None:
            self.point = basectx.point
        else:
            self.point = None

        # Spec
        sp = get_ctx_prop(ctxmap, "spec")
        if isinstance(sp, ProjectNameSpec):
            self.spec = sp
        elif basectx is not None:
            self.spec = basectx.spec
        else:
            self.spec = None

        # Result
        r = get_ctx_prop(ctxmap, "result")
        if isinstance(r, ProjectNameResult):
            self.result = r
        elif basectx is not None:
            self.result = basectx.result
        else:
            self.result = None

        # Response
        rp = get_ctx_prop(ctxmap, "response")
        if isinstance(rp, ProjectNameResponse):
            self.response = rp
        elif basectx is not None:
            self.response = basectx.response
        else:
            self.response = None

        # Resolve operation
        opname = get_ctx_prop(ctxmap, "opname") or ""
        self.op = self.resolve_op(opname)

    def resolve_op(self, opname):
        if opname in self.opmap:
            return self.opmap[opname]

        if opname == "":
            return ProjectNameOperation({})

        entname = "_"
        if self.entity is not None and hasattr(self.entity, "get_name") and callable(self.entity.get_name):
            entname = self.entity.get_name()

        opcfg = vs.getpath(self.config, "entity." + entname + ".op." + opname)

        inpt = "match"
        if opname == "update" or opname == "create":
            inpt = "data"

        points = []
        if isinstance(opcfg, dict):
            t = vs.getprop(opcfg, "points")
            if isinstance(t, list):
                points = t

        op = ProjectNameOperation({
            "entity": entname,
            "name": opname,
            "input": inpt,
            "points": points,
        })

        self.opmap[opname] = op
        return op

    def make_error(self, code, msg):
        return ProjectNameError(code, msg, self)

File: project/.sdk/tm/lua/core/context.lua
-- ProjectName SDK context

local vs = require("utility.struct.struct")
local Control = require("core.control")
local Operation = require("core.operation")
local Spec = require("core.spec")
local Result = require("core.result")
local Response = require("core.response")
local ProjectNameError = require("core.error")
local helpers = require("core.helpers")

local Context = {}
Context.__index = Context


function Context.new(ctxmap, basectx)
  local self = setmetatable({}, Context)

  self.id = "C" .. tostring(math.random(10000000, 99999999))
  self.out = {}

  ctxmap = ctxmap or {}

  -- Client
  local c = helpers.get_ctx_prop(ctxmap, "client")
  if c ~= nil then
    self.client = c
  elseif basectx ~= nil then
    self.client = basectx.client
  end

  -- Utility
  local u = helpers.get_ctx_prop(ctxmap, "utility")
  if u ~= nil then
    self.utility = u
  elseif basectx ~= nil then
    self.utility = basectx.utility
  end

  -- Ctrl
  self.ctrl = Control.new()
  local ctrl_raw = helpers.get_ctx_prop(ctxmap, "ctrl")
  if type(ctrl_raw) == "table" then
    if ctrl_raw.throw_err ~= nil then
      self.ctrl.throw_err = ctrl_raw.throw_err
    elseif type(ctrl_raw["throw"]) == "boolean" then
      self.ctrl.throw_err = ctrl_raw["throw"]
    end
    if type(ctrl_raw.explain) == "table" then
      self.ctrl.explain = ctrl_raw.explain
    end
  elseif basectx ~= nil and basectx.ctrl ~= nil then
    self.ctrl = basectx.ctrl
  end

  -- Meta
  self.meta = {}
  local m = helpers.get_ctx_prop(ctxmap, "meta")
  if type(m) == "table" then
    self.meta = m
  elseif basectx ~= nil and basectx.meta ~= nil then
    self.meta = basectx.meta
  end

  -- Config
  local cfg = helpers.get_ctx_prop(ctxmap, "config")
  if type(cfg) == "table" then
    self.config = cfg
  elseif basectx ~= nil then
    self.config = basectx.config
  end

  -- Entopts
  local eo = helpers.get_ctx_prop(ctxmap, "entopts")
  if type(eo) == "table" then
    self.entopts = eo
  elseif basectx ~= nil then
    self.entopts = basectx.entopts
  end

  -- Options
  local o = helpers.get_ctx_prop(ctxmap, "options")
  if type(o) == "table" then
    self.options = o
  elseif basectx ~= nil then
    self.options = basectx.options
  end

  -- Entity
  local e = helpers.get_ctx_prop(ctxmap, "entity")
  if e ~= nil then
    self.entity = e
  elseif basectx ~= nil then
    self.entity = basectx.entity
  end

  -- Shared
  local s = helpers.get_ctx_prop(ctxmap, "shared")
  if type(s) == "table" then
    self.shared = s
  elseif basectx ~= nil then
    self.shared = basectx.shared
  end

  -- Opmap
  local om = helpers.get_ctx_prop(ctxmap, "opmap")
  if type(om) == "table" then
    self.opmap = om
  elseif basectx ~= nil then
    self.opmap = basectx.opmap
  end
  if self.opmap == nil then
    self.opmap = {}
  end

  -- Data
  self.data = helpers.to_map(helpers.get_ctx_prop(ctxmap, "data")) or {}
  self.reqdata = helpers.to_map(helpers.get_ctx_prop(ctxmap, "reqdata")) or {}
  self.match = helpers.to_map(helpers.get_ctx_prop(ctxmap, "match")) or {}
  self.reqmatch = helpers.to_map(helpers.get_ctx_prop(ctxmap, "reqmatch")) or {}

  -- Point
  local pt = helpers.get_ctx_prop(ctxmap, "point")
  if type(pt) == "table" then
    self.point = pt
  elseif basectx ~= nil then
    self.point = basectx.point
  end

  -- Spec
  local sp = helpers.get_ctx_prop(ctxmap, "spec")
  if sp ~= nil and getmetatable(sp) == Spec then
    self.spec = sp
  elseif basectx ~= nil then
    self.spec = basectx.spec
  end

  -- Result
  local r = helpers.get_ctx_prop(ctxmap, "result")
  if r ~= nil and getmetatable(r) == Result then
    self.result = r
  elseif basectx ~= nil then
    self.result = basectx.result
  end

  -- Response
  local rp = helpers.get_ctx_prop(ctxmap, "response")
  if rp ~= nil and getmetatable(rp) == Response then
    self.response = rp
  elseif basectx ~= nil then
    self.response = basectx.response
  end

  -- Resolve operation
  local opname = helpers.get_ctx_prop(ctxmap, "opname") or ""
  self.op = self:resolve_op(opname)

  return self
end


function Context:resolve_op(opname)
  if self.opmap[opname] ~= nil then
    return self.opmap[opname]
  end

  if opname == "" then
    return Operation.new({})
  end

  local entname = "_"
  if self.entity ~= nil and type(self.entity.get_name) == "function" then
    entname = self.entity:get_name()
  end

  local opcfg = vs.getpath(self.config, "entity." .. entname .. ".op." .. opname)

  local input = "match"
  if opname == "update" or opname == "create" then
    input = "data"
  end

  local points = {}
  if type(opcfg) == "table" then
    local t = vs.getprop(opcfg, "points")
    if type(t) == "table" then
      points = t
    end
  end

  local op = Operation.new({
    entity = entname,
    name = opname,
    input = input,
    points = points,
  })

  self.opmap[opname] = op
  return op
end


function Context:make_error(code, msg)
  return ProjectNameError.new(code, msg, self)
end


return Context
